### ISSUE TITLE
Deadlock in ConsumerWorker in the REST proxy

### ISSUE THREAD

**Issue by rajinisivaram** [confluentinc/kafka-rest#97]
We are regularly seeing deadlocks in the REST proxy when all request threads get blocked waiting for ConsumerWorker object lock. I think there is a potential deadlock since the order of ,locking ConsumerWorker,ConsumerState in the ConsumerWorker thread and the request threads are not always the same. Looking through the stack traces on one of our failures, I think the sequence of events is something like this:

1. Request to read from a consumer group is received. Request is processed and `ConsumerState` is created (let's say `ConsumerStateA`) for the consumer group.
2. Request thread queues a read task
3. `ConsumerWorker` thread (`Thread-9`) sees the read task and calls `doPartialRead()` and waits for results. `Thread-9` now holds a read lock on `ConsumerStateA`. `Thread-9` frees `ConsumerWorker` object lock during the wait.
4. Another request to read from the same consumer group on a different topic is received on thread `qtp-1826533390-54`
5. `ConsumerWorker` lock is free, hence `qtp-1826533390-54` locks `ConsumerWorker` object lock (synchronized block). 
6. `ConsumerState.getOrCreateTopicState()` is invoked for the new topic. The first half of `getOrCreateTopicState()` requires only a read lock, so that is not an issue
7. `getOrCreateTopicState()` now needs a write lock on `ConsumerStateA`. But the write lock cannot be acquired since `ConsumerWorker` (`Thread-9`) is holding a read lock on `ConsumerStateA`

This is a deadlock since `Thread-9` cannot release the read lock until it can acquire the `ConsumerWorker` object lock. And the thread `qtp-1826533390-54` holding that `ConsumerWorker` object lock is blocked on the read lock held by `Thread-9`. Further requests are all blocked on new request processor threads since the `ConsumerWorker` object lock is never released.

The stack traces of relevant threads:

######  ConsumerWorker thread that is holding a read lock and is blocked on the ConsumerWorker lock
```
3XMTHREADINFO      "Thread-9" J9VMThread:0x0000000021CF4000, j9thread_t:0x00007FEBF4CE4170, java/lang/Thread:0x00000000018DD9B0, state:B, prio=5
3XMJAVALTHREAD            (java/lang/Thread getId:0x19, isDaemon:false)
3XMTHREADINFO1            (native thread ID:0x2F2, native priority:0x5, native policy:UNKNOWN, vmstate:B, vm thread flags:0x01080101)
3XMTHREADINFO2            (native stack address range from:0x00007FEBEF89C000, to:0x00007FEBEF8DD000, size:0x41000)
3XMCPUTIME               CPU usage total: 0.420590890 secs
3XMTHREADBLOCK     Blocked on: io/confluent/kafkarest/ConsumerWorker@0x00000000018DD9B0 Owned by: "qtp-1826533390-54 - /consumers/consumer_group/instances/consumer_instance/topics/testtopic24" (J9VMThread:0x0000000021F45100, java/lang/Thread:0x0000000001FB8C60)
3XMHEAPALLOC             Heap bytes allocated since last GC cycle=0 (0x0)
3XMTHREADINFO3           Java callstack:
4XESTACKTRACE                at java/lang/Object.wait(Native Method)
4XESTACKTRACE                at java/lang/Object.wait(Object.java:196)
4XESTACKTRACE                at io/confluent/kafkarest/SystemTime.waitOn(SystemTime.java:22)
4XESTACKTRACE                at io/confluent/kafkarest/ConsumerWorker.run(ConsumerWorker.java:76)
5XESTACKTRACE                   (entered lock: io/confluent/kafkarest/ConsumerWorker@0x00000000018DD9B0, entry count: 1)
```

###### Thread that owns ConsumerWorker lock, and is blocked on the write lock 
```
3XMTHREADINFO      "qtp-1826533390-54 - /consumers/consumer_group/instances/consumer_instance/topics/testtopic24" J9VMThread:0x0000000021F45100, j9thread_t:0x00007FEB90050250, java/lang/Thread:0x00000
00001FB8C60, state:P, prio=5
3XMJAVALTHREAD            (java/lang/Thread getId:0x36, isDaemon:false)
3XMTHREADINFO1            (native thread ID:0x30D, native priority:0x5, native policy:UNKNOWN, vmstate:P, vm thread flags:0x01020001)
3XMTHREADINFO2            (native stack address range from:0x00007FEBEF4CD000, to:0x00007FEBEF50E000, size:0x41000)
3XMCPUTIME               CPU usage total: 0.662405516 secs
3XMTHREADBLOCK     Parked on: java/util/concurrent/locks/ReentrantReadWriteLock$NonfairSync@0x000000000244F610 Owned by: <unknown>
3XMHEAPALLOC             Heap bytes allocated since last GC cycle=0 (0x0)
3XMTHREADINFO3           Java callstack:
4XESTACKTRACE                at sun/misc/Unsafe.park(Native Method)
4XESTACKTRACE                at java/util/concurrent/locks/LockSupport.park(LockSupport.java:198(Compiled Code))
4XESTACKTRACE                at java/util/concurrent/locks/AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:846(Compiled Code))
4XESTACKTRACE                at java/util/concurrent/locks/AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:879(Compiled Code))
4XESTACKTRACE                at java/util/concurrent/locks/AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1209(Compiled Code))
4XESTACKTRACE                at java/util/concurrent/locks/ReentrantReadWriteLock$WriteLock.lock(ReentrantReadWriteLock.java:957(Compiled Code))
4XESTACKTRACE                at io/confluent/kafkarest/ConsumerState.getOrCreateTopicState(ConsumerState.java:190)
4XESTACKTRACE                at io/confluent/kafkarest/ConsumerReadTask.<init>(ConsumerReadTask.java:77)
4XESTACKTRACE                at io/confluent/kafkarest/ConsumerWorker.readTopic(ConsumerWorker.java:54)
5XESTACKTRACE                   (entered lock: io/confluent/kafkarest/ConsumerWorker@0x00000000018DD9B0, entry count: 1)
4XESTACKTRACE                at io/confluent/kafkarest/ConsumerManager.readTopic(ConsumerManager.java:247)
4XESTACKTRACE                at com/ibm/mariposa/kafkarest/resources/MConsumersResource.readTopic(MConsumersResource.java:159)
4XESTACKTRACE                at com/ibm/mariposa/kafkarest/resources/MConsumersResource.readTopicBinary(MConsumersResource.java:136)
4XESTACKTRACE                at sun/reflect/GeneratedMethodAccessor15.invoke(Bytecode PC:175)
4XESTACKTRACE                at sun/reflect/DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:56(Compiled Code))
4XESTACKTRACE                at java/lang/reflect/Method.invoke(Method.java:620(Compiled Code))
4XESTACKTRACE                at org/glassfish/jersey/server/model/internal/ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
4XESTACKTRACE                at org/glassfish/jersey/server/model/internal/AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:151)
4XESTACKTRACE                at org/glassfish/jersey/server/model/internal/AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:171)
4XESTACKTRACE                at org/glassfish/jersey/server/model/internal/JavaResourceMethodDispatcherProvider$VoidOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:136)
4XESTACKTRACE                at org/glassfish/jersey/server/model/internal/AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:104)
4XESTACKTRACE                at org/glassfish/jersey/server/model/ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:406)
4XESTACKTRACE                at org/glassfish/jersey/server/model/ResourceMethodInvoker.apply(ResourceMethodInvoker.java:350)
4XESTACKTRACE                at org/glassfish/jersey/server/model/ResourceMethodInvoker.apply(ResourceMethodInvoker.java:106)
4XESTACKTRACE                at org/glassfish/jersey/server/ServerRuntime$1.run(ServerRuntime.java:259)
```

######  Other request processor threads blocked on ConsumerWorker object lock
```
3XMTHREADINFO      "qtp-1826533390-36 - /consumers/consumer_group/instances/consumer_instance/topics/testtopic23" J9VMThread:0x0000000021F52200, j9thread_t:0x00007FEBF512A440, java/lang/Thread:0x0000000001B045C8, state:B, prio=5
3XMJAVALTHREAD            (java/lang/Thread getId:0x24, isDaemon:false)
3XMTHREADINFO1            (native thread ID:0x2FB, native priority:0x5, native policy:UNKNOWN, vmstate:B, vm thread flags:0x01000201)
3XMTHREADINFO2            (native stack address range from:0x00007FEBEF653000, to:0x00007FEBEF694000, size:0x41000)
3XMCPUTIME               CPU usage total: 0.959511346 secs
3XMTHREADBLOCK     Blocked on: io/confluent/kafkarest/ConsumerWorker@0x00000000018DD9B0 Owned by: "qtp-1826533390-54 - /consumers/consumer_group/instances/consumer_instance/topics/testtopic24" (J9VMThread:0x0000000021F45100, java/lang/Thread:0x0000000001FB8C60)
3XMHEAPALLOC             Heap bytes allocated since last GC cycle=0 (0x0)
3XMTHREADINFO3           Java callstack:
4XESTACKTRACE                at io/confluent/kafkarest/ConsumerManager.readTopic(ConsumerManager.java:247)
4XESTACKTRACE                at com/ibm/mariposa/kafkarest/resources/MConsumersResource.readTopic(MConsumersResource.java:159)
4XESTACKTRACE                at com/ibm/mariposa/kafkarest/resources/MConsumersResource.readTopicBinary(MConsumersResource.java:136)
4XESTACKTRACE                at sun/reflect/GeneratedMethodAccessor15.invoke(Bytecode PC:175)
4XESTACKTRACE                at sun/reflect/DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:56(Compiled Code))
```

**Comment by johnstark** [comment]
@rajinisivaram , did you resolve this issue?  If so, how?  Thanks.

**Comment by rajinisivaram** [comment]
@johnstark We fixed this in our REST proxy by disallowing multiple concurrent operations for the same consumer instance. We added a per-consumer-instance semaphore. If a request from a consumer instance is received while a previous request is in progress, we return error code 409 and don't call into the consumer manager. This avoids the deadlock. Changing the locking in the consumer manager to avoid deadlocking looked a bit complicated at the time, so we went for this simpler solution.
